In [ ]:
import os

def rename_jpg_files_to_numbers(directory):
    files = os.listdir(directory)
    number = 1

    for file in files:
        file_path = os.path.join(directory, file)
        if os.path.isfile(file_path) and file.lower().endswith('.png'):
            new_name = f"{number}.jpg"
            new_file_path = os.path.join(directory, new_name)
            os.rename(file_path, new_file_path)
            number += 1

directory_path = '/content/obfuscated_private_ed/2023-04-11-09-18-38-day/back_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-11-09-18-38-day/front_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-11-09-18-38-day/masks/back_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-11-09-18-38-day/masks/front_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-13-13-07-34-night/back_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-13-13-07-34-night/front_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-13-13-07-34-night/masks/back_cam'
rename_jpg_files_to_numbers(directory_path)
directory_path = '/content/obfuscated_private_ed/2023-04-13-13-07-34-night/masks/front_cam'
rename_jpg_files_to_numbers(directory_path)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

## Установка в Google Colab

Ячейки ниже рекомендуется использовать для установки зависимостей в Google Colab.

1. Убедитесь, что вы подключены к окружению с GPU

In [ ]:
!nvidia-smi

Thu Jul  4 07:06:25 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   47C    P8              11W /  70W |      0MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

2. Установка необходимых библиотек для сборки MinkowskiEngine

In [ ]:
!sudo apt-get install libopenblas-dev

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libopenblas-dev is already the newest version (0.3.20+ds-1).
0 upgraded, 0 newly installed, 0 to remove and 45 not upgraded.


In [ ]:
!pip install ninja

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 307.2/307.2 kB 6.0 MB/s eta 0:00:00


3. Сборка и установка MinkowskiEngine из репозитория для работы с CUDA 12.2 (занимает много времени)

In [ ]:
!pip install -U git+https://github.com/richlukich/MinkowskiEngine -v --no-deps \
                          --config-settings="--install-option=--force_cuda" \
                          --config-settings="--install-option=--blas=openblas"

Using pip 23.1.2 from /usr/local/lib/python3.10/dist-packages/pip (python 3.10)
  Cloning https://github.com/richlukich/MinkowskiEngine to /tmp/pip-req-build-c3dcs7ws
  Running command git version
  git version 2.34.1
  Running command git clone --filter=blob:none https://github.com/richlukich/MinkowskiEngine /tmp/pip-req-build-c3dcs7ws
  Cloning into '/tmp/pip-req-build-c3dcs7ws'...
  Updating files:   0% (2/244)
  Updating files:   1% (3/244)
  Updating files:   2% (5/244)
  Updating files:   3% (8/244)
  Updating files:   4% (10/244)
  Updating files:   5% (13/244)
  Updating files:   6% (15/244)
  Updating files:   7% (18/244)
  Updating files:   8% (20/244)
  Updating files:   9% (22/244)
  Updating files:  10% (25/244)
  Updating files:  11% (27/244)
  Updating files:  12% (30/244)
  Updating files:  13% (32/244)
  Updating files:  14% (35/244)
  Updating files:  15% (37/244)
  Updating files:  16% (40/244)
  Updating files:  17% (42/244)
  Updating files:  18% (44/244)
  Updatin

4. Проверка, что все работает

In [ ]:
import torch
print(f"Is CUDA available in torch?: {torch.cuda.is_available()}")
import MinkowskiEngine as ME
print(f"Is CUDA available in MinkowskiEngine?: {ME.is_cuda_available()}")
ME.print_diagnostics()

Is CUDA available in torch?: True
Is CUDA available in MinkowskiEngine?: True
==========System==========
Linux-6.1.85+-x86_64-with-glibc2.35
3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
==========Pytorch==========
2.3.0+cu121
torch.cuda.is_available(): True
==========NVIDIA-SMI==========
Driver Version 535.104.05
CUDA Version 12.2
VBIOS Version 90.04.A7.00.01
Image Version G183.0200.00.02
GSP Firmware Version N/A
==========NVCC==========
==========CC==========
==========MinkowskiEngine==========
0.5.4
MinkowskiEngine compiled with CUDA Support: True
NVCC version MinkowskiEngine is compiled: 12020
CUDART version MinkowskiEngine is compiled: 12020


5. Финальный шаг - установка библиотеки [opr](https://github.com/alexmelekhin/open_place_recognition), код из которой будет использоваться в бейзлайне

In [ ]:
!git clone -b feat/itlp_outdoor https://github.com/alexmelekhin/open_place_recognition
%cd open_place_recognition
!pip install -e .  # флаг -e необходим для возможности редактировать код уже установленной библиотеки
%cd ..

Cloning into 'open_place_recognition'...
remote: Enumerating objects: 1467, done.
remote: Counting objects: 100% (696/696), done.
remote: Compressing objects: 100% (304/304), done.
remote: Total 1467 (delta 470), reused 494 (delta 376), pack-reused 771
Receiving objects: 100% (1467/1467), 22.87 MiB | 10.29 MiB/s, done.
Resolving deltas: 100% (842/842), done.
/content/open_place_recognition
Obtaining file:///content/open_place_recognition
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.5/154.5 kB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 7.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.1/119.1 kB 15.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 19.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━

/content


In [ ]:
!pip install loguru

In [ ]:
!python -m pip install numpy-quaternion

## Загрузка датасета в Google Colab

Пример кода для загрузки датасета.

Вы можете воспользоваться утилитой gdown, которая по умолчанию доступна в Colab. Допустим, https://drive.google.com/file/d/1rXsx0GEotwwTHjX1PveKdGzcatssO7wg/view?usp=sharing - ссылка на файл. Чтобы скачать его, нам нужно передать в gdown в качестве аргумента его id - для данного примера это `1rXsx0GEotwwTHjX1PveKdGzcatssO7wg` (часть ссылки между `file/d/` и `/view`).

In [ ]:
#!gdown 1txG6aPiy5XtxLOxFk0CxFk9Ey_VUsB5T

Вы можете сверить хэш-сумму файла:

In [ ]:
#!sha256sum public.zip

И распаковать архив:

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
!unzip -q '/content/drive/MyDrive/Colab Notebooks/public.zip'

## Базовое решение

Создаем даталоадеры

In [ ]:
!unzip -q '/content/drive/MyDrive/Colab Notebooks/obfuscated_private_ed.zip'

replace __MACOSX/._obfuscated_private_ed? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [ ]:
DATA_ROOT = "/content/public"

In [ ]:
DATA_ROOT2 = "/content/obfuscated_private_ed"

In [ ]:
from torch.utils.data import DataLoader

from opr.datasets.itlp_outdoor import ITLPCampusOutdoor
from opr.samplers import BatchSampler

In [ ]:
train_dataset = ITLPCampusOutdoor(
    DATA_ROOT,
    subset="train",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)
val_dataset = ITLPCampusOutdoor(
    DATA_ROOT,
    subset="val",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)

test_dataset = ITLPCampusOutdoor(
    DATA_ROOT,
    subset="test",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)

In [ ]:
train_sampler = BatchSampler(
    train_dataset,
    batch_size=16,  # initial batch size
    batch_size_limit=256,  # maximum batch size (see "Dynamic Batch Sizing")
    batch_expansion_rate=1.4,
    drop_last=True,
)
val_sampler = BatchSampler(
    val_dataset,
    batch_size=256,  # initial batch size
    drop_last=True,
)

In [ ]:
train_dl = DataLoader(
    dataset=train_dataset,
    batch_sampler=train_sampler,
    collate_fn=train_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
)
val_dl = DataLoader(
    dataset=val_dataset,
    batch_sampler=val_sampler,
    collate_fn=val_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
)

test_dl = DataLoader(
    dataset=test_dataset,
    batch_size=256,
    collate_fn=test_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
    drop_last=False,
)


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:558: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


## BLIP + BERT

### BLIP

In [ ]:
!pip3 install salesforce-lavis

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 28.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.6/13.6 MB 59.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.2/43.2 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.4/235.4 kB 25.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.4/54.4 kB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.2/42.2 kB 6.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.8/47.8 MB 10.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 204.3/204.3 kB 29.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.3/104.3 MB 9.4 MB/s eta 0:00:00
     ━━━━━━

In [ ]:
import torch
from PIL import Image

from lavis.models import load_model_and_preprocess

In [ ]:
device = torch.device("cuda") if torch.cuda.is_available() else "cpu"

### текстовые данные для модели BERT

## В качестве базовой модели будем использовать ResNet18FPN для извлечения признаков изображения, Generalized Mean (GeM) pooling, фьюзить данные будем с помощью суммирования

In [ ]:
import torch

from opr.models.place_recognition.base import ImageModel
from opr.modules.feature_extractors import ResNet18FPNFeatureExtractor
from opr.modules import Add, GeM, Concat

In [ ]:
from typing import Dict, Optional

from loguru import logger
from torch import Tensor, nn
import torch.nn.functional as F

In [ ]:
class ImageModel2(nn.Module):
    """Meta-model for image-based Place Recognition. Combines feature extraction backbone and head modules."""

    def __init__(
        self,
        backbone: nn.Module,
        head: nn.Module,
        image2text,
        text_embedding,
        tokenizer,
        fusion: Optional[nn.Module] = None,
    ) -> None:
        """Meta-model for image-based Place Recognition.

        Args:
            backbone (ImageFeatureExtractor): Image feature extraction backbone.
            head (ImageHead): Image head module.
            fusion (FusionModule, optional): Module to fuse descriptors for multiple images in batch.
                Defaults to None.
        """
        super().__init__()
        self.backbone = backbone
        self.head = head
        self.fusion = fusion
        self.image2text = image2text
        self.text_embedding = text_embedding
        self.tokenizer = tokenizer

        self.linear = nn.Linear(1024, 256)
        self.batch_norm = nn.BatchNorm1d(256)
        #self.linear_layer1 = nn.Linear(, 256)

    def get_describe(self, image_tensor):
        res = image_tensor.unsqueeze(0)
        resized_images = F.interpolate(res, size=(384, 384), mode='bilinear', align_corners=False).to(device)
        return self.image2text.generate({"image": resized_images})


    def forward(self, batch: Dict[str, Tensor]) -> Dict[str, Tensor]:  # noqa: D102
        img_descriptors = {}
        for key, value in batch.items():
            if key.startswith("images_"):
                text = [self.get_describe(b)[0] for b in value]
                encoded_input = self.tokenizer(text , padding=True, truncation=True, return_tensors='pt').to(device)
                model_output = self.text_embedding(**encoded_input)
                cls_embedding = model_output.last_hidden_state[:, 0, :]
                images_embedding = self.head(self.backbone(value))
                print(images_embedding.shape, cls_embedding.shape)
                concatenated = torch.cat((images_embedding, cls_embedding), dim=1)
                compressed = self.linear(concatenated)
                activated = F.relu(compressed)
                normalized = self.batch_norm(activated)
                result =  images_embedding + normalized
                img_descriptors[key] = result

        if len(img_descriptors) > 1:
            if self.fusion is None:
                raise ValueError("Fusion module is not defined but multiple images are provided")
            descriptor = self.fusion(img_descriptors)
        else:
            if self.fusion is not None:
                raise ValueError("Fusion module is defined but only one image is provided")
            descriptor = list(img_descriptors.values())[0]
        out_dict: Dict[str, Tensor] = {"final_descriptor": descriptor}
        return out_dict

In [ ]:
from transformers import BertTokenizer, BertModel
import torch

In [ ]:
from opr.modules.feature_extractors import ResNet50FPNFeatureExtractor

In [ ]:
feature_extractor = ResNet50FPNFeatureExtractor(
    in_channels=3,
    lateral_dim=256,
    fh_num_bottom_up=4,
    fh_num_top_down=0,
    pretrained=True,
)
pooling = GeM()
descriptor_fusion_module = Concat()


In [ ]:
model = ImageModel2(
    backbone=feature_extractor,
    head=pooling,
    image2text=blip,
    text_embedding = bert,
    tokenizer=tokenizer,
    fusion=descriptor_fusion_module,
)

In [ ]:
model = ImageModel(backbone = feature_extractor, head = pooling, fusion = descriptor_fusion_module)

In [ ]:
IMAGE_LR = 0.0001
HEAD_LR = 0.0001
FUSION_LR = 0.0001
WEIGHT_DECAY = 0.0001
SCHEDULER_GAMMA = 0.1
SCHEDULER_STEPS = [5]

In [ ]:
from torch.optim import Adam
from torch.optim.lr_scheduler import MultiStepLR

In [ ]:
params_list = []
if model.backbone is not None and IMAGE_LR is not None:
    params_list.append({"params": model.backbone.parameters(), "lr": IMAGE_LR})
if model.head is not None and HEAD_LR is not None:
    params_list.append({"params": model.head.parameters(), "lr": HEAD_LR})
if model.fusion is not None and FUSION_LR is not None:
    params_list.append({"params": model.fusion.parameters(), "lr": FUSION_LR})

In [ ]:
from opr.trainers.place_recognition import UnimodalPlaceRecognitionTrainer
from opr.losses import BatchHardTripletMarginLoss

In [ ]:
!mkdir checkpoints

mkdir: cannot create directory ‘checkpoints’: File exists


Инициализируем наш Trainer

In [ ]:
loss_fn = BatchHardTripletMarginLoss(margin=0.2)
optimizer = Adam(params_list, weight_decay=WEIGHT_DECAY)
scheduler = MultiStepLR(optimizer, milestones=SCHEDULER_STEPS, gamma=SCHEDULER_GAMMA)

trainer = UnimodalPlaceRecognitionTrainer(
    checkpoints_dir='/content/checkpoints',
    model=model,
    loss_fn=loss_fn,
    optimizer=optimizer,
    scheduler=scheduler,
    batch_expansion_threshold=0.7,
    wandb_log=False,
    device='cuda',
)

Обучаем

In [ ]:
trainer.train(
    epochs=2,
    train_dataloader=train_dl,
    val_dataloader=val_dl,
    test_dataloader=test_dl,
)

2024-07-04 08:42:03.117 | INFO     | opr.trainers.place_recognition.unimodal:train:113 - =====> Epoch:   1/2:
2024-07-04 08:42:03.119 | INFO     | opr.trainers.place_recognition.unimodal:_loop_epoch:244 - => Train stage:
Train:   0%|          | 0/88 [00:00<?, ?it/s]/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
Train: 100%|██████████| 88/88 [01:31<00:00,  1.90it/s]/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
2024-07-04 08:43:34.949 | INFO     | opr.trainers.place_recognition.unimodal:_loop_epoch:277 - Train time: 01:31
2024-07-04 08:43:34.952 | INFO     | opr.trainers.place_recognition.unimodal:_loop_epoch:278 - Train st

Cтоит отметить, что при обучении тестирование происходит с помощью параметра distance_threshold = 25, подсчет метрик в яндекс контесте происходит при параметре distance_threshold = 5, поэтому делаем перерасчет

In [ ]:
trainer.test(test_dl, distance_threshold=5)

In [ ]:
DATA_ROOT2 = "/content/obfuscated_private_ed"

In [ ]:
train_dataset = ITLPCampusOutdoor(
    DATA_ROOT2,
    subset="train",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)
val_dataset = ITLPCampusOutdoor(
    DATA_ROOT2,
    subset="val",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)

test_dataset = ITLPCampusOutdoor(
    DATA_ROOT2,
    subset="test",
    sensors=("front_cam", "back_cam"),
    load_semantics=True,
)

In [ ]:
train_sampler = BatchSampler(
    train_dataset,
    batch_size=16,  # initial batch size
    batch_size_limit=256,  # maximum batch size (see "Dynamic Batch Sizing")
    batch_expansion_rate=1.4,
    drop_last=True,
)
val_sampler = BatchSampler(
    val_dataset,
    batch_size=256,  # initial batch size
    drop_last=True,
)

In [ ]:
train_dl = DataLoader(
    dataset=train_dataset,
    batch_sampler=train_sampler,
    collate_fn=train_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
)
val_dl = DataLoader(
    dataset=val_dataset,
    batch_sampler=val_sampler,
    collate_fn=val_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
)

test_dl = DataLoader(
    dataset=test_dataset,
    batch_size=256,
    collate_fn=test_dataset.collate_fn,
    num_workers=4,
    pin_memory=True,
    drop_last=False,
)

/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:558: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


## Подготовка ответа для загрузки на сервер

Пример кода для создания файла сабмита. Вы можете модифицировать пайплайн, например добавить ре-ранжирование кандидатов на основе каких-либо характеристик (например, как в [Path-NetVLAD](https://arxiv.org/abs/2103.01486)).

In [ ]:
import itertools

import pandas as pd
from sklearn.neighbors import KDTree
import numpy as np
import torch
from tqdm import tqdm


def extract_embeddings(model, descriptor_key, dataloader, device):
    model = model.to(device)
    model.eval()
    with torch.no_grad():
        test_embeddings_list = []
        for data in tqdm(dataloader, desc="Calculating test set descriptors"):
            batch = data
            batch = {e: batch[e].to(device) for e in batch}
            batch_embeddings = model(batch)
            test_embeddings_list.append(batch_embeddings[descriptor_key].cpu().numpy())
        test_embeddings = np.vstack(test_embeddings_list)
    return test_embeddings


def test_submission(
    test_embeddings: np.ndarray, dataset_df: pd.DataFrame, filename: str = "submission.txt"
) -> None:
    """Function to create submission txt file.

    Args:
        test_embeddings (np.ndarray): Array of embeddings.
        dataset_df (pd.Dataframe): Test dataset dataframe ('test.csv').
        filename (str): Name of the output txt file. Defaults to "submission.txt".
    """
    tracks = []

    for _, group in dataset_df.groupby("track"):
        tracks.append(group.index.to_numpy())
    n = 1
    ij_permutations = sorted(list(itertools.permutations(range(len(tracks)), 2)))
    # ij_permutations = [(0, 1), (0, 2), (1, 0), (1, 2), (2, 0), (2, 1)]

    submission_lines = []

    for i, j in tqdm(ij_permutations, desc="Calculating metrics"):
        query_indices = tracks[i]
        database_indices = tracks[j]
        query_embs = test_embeddings[query_indices]
        database_embs = test_embeddings[database_indices]

        database_tree = KDTree(database_embs)
        _, indices = database_tree.query(query_embs, k=n)

        submission_lines.extend(list(database_indices[indices.squeeze()]))

    with open(filename, "w") as f:
        for l in submission_lines:
            f.write(str(l)+"\n")

In [ ]:
embeddings = extract_embeddings(model, descriptor_key="final_descriptor", dataloader=test_dl, device='cuda')
test_submission(embeddings, dataset_df=test_dl.dataset.dataset_df, filename="baseline_submission.txt")

Calculating test set descriptors:   0%|          | 0/5 [00:00<?, ?it/s]/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
Calculating test set descriptors:   0%|          | 0/5 [00:00<?, ?it/s]


ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)  # type: ignore[possibly-undefined]
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/content/open_place_recognition/src/opr/datasets/itlp_outdoor.py", line 469, in __getitem__
    im = self._load_image("front_cam", idx, track, floor)
  File "/content/open_place_recognition/src/opr/datasets/itlp_outdoor.py", line 268, in _load_image
    image_ts = int(self.dataset_df[f"{cam}_ts"].iloc[idx])
ValueError: invalid literal for int() with base 10: '0a4038a6fbed52dbd04ae6d3ffc2d43d527bf6f1ed530a0c256f093cb3409229'


## Текстовая информация

Для решения задачи также предлагается использовать OCR-пайплайн, [реализация](https://colab.research.google.com/drive/1TXGPCjOdi7auAIoHOspBqS1Tr-c9hHym?usp=sharing) которого представлена в библиотеке OpenPlaceRecognition. Таким образом, для предсказания места помимо текстовых эмбеддингов можно использовать текст, найденный на сцене.

Файл с сабмитом необходимо загружать на яндекс контест: https://contest.yandex.ru/contest/63631

Удачи!